### PR TITLE
Access to previous document in columnstore2

### PR DESCRIPTION
**Pull request by gnusi**
Allow accessing previous document from columnstore2 iterators.

### CODE REVIEW

File: core/utils/attribute_helper.hpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+struct type<attribute_ptr<T>> : type<T> {};
+
+namespace detail {
+
+template<size_t I, typename... T>
+constexpr attribute_ptr<attribute> get_mutable_helper(
+    std::tuple<T...>& t, type_info::type_id id) noexcept {
+  auto& v = std::get<I>(t);
+  if (type<std::remove_reference_t<decltype(v)>>::id() == id) {
+    return v;
```
Comment (MBkkt) [inline core/utils/attribute_helper.hpp:58]: Why don't you return `static_cast<attribute*>(&v);` instead of `attribute_ptr<attribute>`?

Comment (MBkkt) [inline core/utils/attribute_helper.hpp:58]:
```suggestion
  if (type<std::tuple_element_t<I, std::tuple<T...>>>::id() == id) {
    return std::get<I>(t);
```

File: core/formats/columnstore2.cpp [hunk 2]
```diff
@@ -157,33 +151,36 @@ std::vector<uint64_t> read_blocks_dense(
...
+        min_base_{header.min},
+        min_doc_{min_base_},
+        max_doc_{min_base_ + header.docs_count - 1} {
     assert(min_doc_ <= max_doc_);
     assert(!doc_limits::eof(max_doc_));
-    std::get<irs::cost>(attrs_).reset(header.docs_count);
+    std::get<cost>(attrs_).reset(header.docs_count);
+    std::get<seek_prev>(attrs_).reset(
+        [](const void* ctx) {
+          auto* self = reinterpret_cast<const range_column_iterator*>(ctx);
```
Comment (MBkkt) [inline core/formats/columnstore2.cpp:176]:
```suggestion
          auto* self = static_cast<const range_column_iterator*>(ctx);
```

File: core/formats/sparse_bitmap.cpp [hunk 3]
```diff
@@ -269,23 +288,25 @@ struct container_iterator<BT_DENSE> {
...
+
       return true;
     }
 
-    constexpr int32_t NUM_BLOCKS = sparse_bitmap_writer::kNumBlocks;
+    constexpr int32_t kNumBlocks = sparse_bitmap_writer::kNumBlocks;
 
     ++ctx.word_idx;
-    for (; ctx.word_idx < NUM_BLOCKS; ++ctx.word_idx) {
+    for (; ctx.word_idx < kNumBlocks; ++ctx.word_idx) {
```
Comment (MBkkt) [inline core/formats/sparse_bitmap.cpp:301]: NIT
It's used only once, maybe better to remove constant

File: core/formats/sparse_bitmap.cpp [hunk 4]
```diff
@@ -310,9 +332,84 @@ struct container_iterator<BT_DENSE> {
...
+          --word_idx;
+          prev -= sizeof(size_t);
+        } while (!word);
+        self->in_->seek(pos);
+      } else {
+        for (auto* prev_word = ctx.u64data - 1; !word; --word_idx) {
+          word = *--prev_word;
+          assert(word_idx);
+        }
+        if constexpr (!is_big_endian()) {
```
Comment (MBkkt) [inline core/formats/sparse_bitmap.cpp:385]: Maybe better in ntoh?

File: core/formats/sparse_bitmap.cpp [hunk 5]
```diff
@@ -269,23 +288,25 @@ struct container_iterator<BT_DENSE> {
     if (left) {
       const doc_id_t offset = std::countr_zero(left);
       std::get<document>(self->attrs_).value = target + offset;
-      std::get<value_index>(self->attrs_).value
-        = ctx.popcnt - std::popcount(left);
+
+      auto& index = std::get<value_index>(self->attrs_).value;
+      index = ctx.popcnt - std::popcount(left);
```
Comment (MBkkt) [inline core/formats/sparse_bitmap.cpp:293]: I thought we don't have `std::popcount` on macOS

**Review by MBkkt — APPROVED** [review]

File: core/utils/attribute_helper.hpp (hunk 1 above)
Comment (MBkkt) [reply core/utils/attribute_helper.hpp:58]: Also you don't need pass `std::tuple<T...>&` instead of `Tuple&`

**Review by Dronplane — APPROVED** [review]
LGTM

File: core/formats/columnstore.cpp [hunk 6]
```diff
@@ -2011,8 +2011,8 @@ class sparse_column final : public column {
     return memory::make_managed<iterator_t>(
         *this,
         refs_.data(),
-        refs_.data() + refs_.size() - 1,
-        hint == ColumnHint::kNormal); // -1 for upper bound
+        refs_.data() + refs_.size() - 1, // -1 for upper bound
+        ColumnHint::kNormal == (hint & ColumnHint::kConsolidation));
```
Comment (MBkkt) [inline core/formats/columnstore.cpp:2015]: NIT
I don't like how it look
Maybe something like this
```suggestion
        (hint & ColumnHint::kConsolidation) != ColumnHint::kConsolidation);
```

File: core/formats/columnstore.cpp [hunk 7]
```diff
@@ -2158,7 +2158,7 @@ class dense_fixed_offset_column final : public column {
 
     return memory::make_managed<iterator_t>(
         *this, refs_.data(), refs_.data() + refs_.size(),
-        hint == ColumnHint::kNormal);
+        ColumnHint::kNormal == (hint & ColumnHint::kConsolidation));
```
Comment (MBkkt) [inline core/formats/columnstore.cpp:2161]: same here

File: core/formats/columnstore2.cpp [hunk 8]
```diff
@@ -157,33 +166,52 @@ std::vector<uint64_t> read_blocks_dense(
...
+        min_doc_{min_base_},
+        max_doc_{min_base_ + header.docs_count - 1} {
     assert(min_doc_ <= max_doc_);
     assert(!doc_limits::eof(max_doc_));
-    std::get<irs::cost>(attrs_).reset(header.docs_count);
+    std::get<cost>(attrs_).reset(header.docs_count);
+    if (track_prev) {
+      std::get<seek_prev>(attrs_).reset(
+          [](const void* ctx) noexcept {
+            auto* self = reinterpret_cast<const range_column_iterator*>(ctx);
```
Comment (MBkkt) [inline core/formats/columnstore2.cpp:195]:
```suggestion
            auto* self = static_cast<const range_column_iterator*>(ctx);
```

File: core/formats/columnstore2.cpp [hunk 9]
```diff
@@ -1279,18 +1237,16 @@ columnstore_writer::column_t writer::push_column(
...
+      column::context{.alloc = alloc_,
+                      .data_out = data_out_.get(),
+                      .cipher = cipher,
+                      .u8buf = buf_.get(),
+                      .consolidation = consolidation_,
+                      .version = ToSparseBitmapVersion(info)},
+      static_cast<field_id>(id), compression, std::move(finalizer),
+      std::move(compressor));
+
+  return std::make_pair(id, [&column](doc_id_t doc) -> column_output& {
```
Comment (MBkkt) [inline core/formats/columnstore2.cpp:1249]:
```suggestion
  return {id, [&column](doc_id_t doc) -> column_output& {
```

File: core/formats/sparse_bitmap.cpp [hunk 10]
```diff
@@ -328,15 +421,24 @@ struct container_iterator<BT_DENSE> {
...
+      origin_{cont_begin_},
+      use_block_index_{opts.use_block_index},
+      prev_doc_written_{opts.version >= SparseBitmapVersion::kPrevDoc},
+      track_prev_doc_{prev_doc_written_ && opts.track_prev_doc} {
   assert(in_);
+
+  if (track_prev_doc_) {
+    std::get<seek_prev>(attrs_).reset(
+        [](const void* ctx) noexcept {
+          return *reinterpret_cast<const doc_id_t*>(ctx);
```
Comment (MBkkt) [inline core/formats/sparse_bitmap.cpp:438]:
```suggestion
          return *static_cast<const doc_id_t*>(ctx);
```

File: core/formats/sparse_bitmap.cpp [hunk 11]
```diff
@@ -352,48 +454,55 @@ void sparse_bitmap_iterator::read_block_header() {
...
   index_ = index_max_;
   index_max_ += popcnt;
+  if (prev_doc_written_) {
+    prev_ = in_->read_int();  // last doc in a previously filled block
+  }
+
+  if (track_prev_doc_) {
+    std::get<seek_prev>(attrs_).reset(
+        [](const void* ctx) noexcept {
+          return *reinterpret_cast<const doc_id_t*>(ctx);
```
Comment (MBkkt) [inline core/formats/sparse_bitmap.cpp:464]:
```suggestion
          return *static_cast<const doc_id_t*>(ctx);
```

File: core/formats/columnstore.cpp (hunk 6 above)
Comment (Dronplane) [inline core/formats/columnstore.cpp:2015]: why not just?
```suggestion
        !(hint & ColumnHint::kConsolidation));
```

File: core/formats/sparse_bitmap.cpp [hunk 12]
```diff
@@ -352,48 +454,55 @@ void sparse_bitmap_iterator::read_block_header() {
...
-      ? &container_iterator<type>::seek<AT_DIRECT>
-      : &container_iterator<type>::seek<AT_STREAM>;
+    ctx_.sparse.index = index_;
+
+    ResolveType(
+        [this]<AccessType A, bool TrackPrev>() noexcept {
+          seek_func_ =
+              &container_iterator<BT_SPARSE, TrackPrev>::template seek<A>;
+        },
+        ctx_.u8data, track_prev_doc_);
```
Comment (MBkkt) [inline core/formats/sparse_bitmap.cpp:486]: make for sparse_bitmap_iterator function: set_seek_func
```suggestion
    set_seek_func<BT_DENSE>();
```

set_seek_func:
```C++
template <BlockType Type>
void sparse_bitmap_iterator::set_seek_func() {
  auto impl = [&]<bool TrackPrev>() {
    if (ctx_.u8data) {
      return container_iterator<Type, TrackPrev>::template seek<AT_DIRECT>;
    } else {
      return container_iterator<Type, TrackPrev>::template seek<AT_STREAM>;
    }
  };

  // FIXME(gnusi): check alignment
  if (track_prev_doc_) {
    seek_func_ = &impl.template operator()<true>();
  } else {
    seek_func_ = &impl.template operator()<false>();
  }
}
```

File: core/formats/columnstore.cpp (hunk 7 above)
Comment (Dronplane) [inline core/formats/columnstore.cpp:2161]: same as above

File: core/formats/sparse_bitmap.cpp [hunk 13]
```diff
@@ -403,10 +512,12 @@ void sparse_bitmap_iterator::read_block_header() {
...
-    // FIXME check alignment
-    seek_func_ = ctx_.u8data
-      ? &container_iterator<type>::seek<AT_DIRECT>
-      : &container_iterator<type>::seek<AT_STREAM>;
+    ResolveType(
+        [this]<AccessType A, bool TrackPrev>() noexcept {
+          seek_func_ =
+              &container_iterator<BT_DENSE, TrackPrev>::template seek<A>;
+        },
+        ctx_.u8data, track_prev_doc_);
```
Comment (MBkkt) [inline core/formats/sparse_bitmap.cpp:520]: make for sparse_bitmap_iterator function: set_seek_func
```suggestion
    set_seek_func<BT_DENSE>();
```

File: core/formats/sparse_bitmap.cpp [hunk 14]
```diff
@@ -28,60 +28,43 @@ namespace {
...
-  //////////////////////////////////////////////////////////////////////////////
+  // Dense block is represented as a bitset container
   BT_DENSE = 0,
 
-  //////////////////////////////////////////////////////////////////////////////
-  /// @brief sprase block is represented as an array of values
-  //////////////////////////////////////////////////////////////////////////////
-  BT_SPARSE
-}; // BlockType
+  // Sprase block is represented as an array of values
```
Comment (Dronplane) [inline core/formats/sparse_bitmap.cpp:43]:
```suggestion
  // Sparse block is represented as an array of values
```

File: core/formats/sparse_bitmap.hpp [hunk 15]
```diff
@@ -145,58 +154,89 @@ class sparse_bitmap_writer {
...
+    return !static_cast<bool>(*this);
   }
-}; // value_index
 
-////////////////////////////////////////////////////////////////////////////////
-/// @class sparse_bitmap_iterator
-////////////////////////////////////////////////////////////////////////////////
+  constexpr bool operator!=(std::nullptr_t) const noexcept {
+    return !(*this == nullptr);
+  }
```
Comment (MBkkt) [inline core/formats/sparse_bitmap.hpp:184]: Since C++20 no needed?
image

File: core/formats/columnstore.cpp (hunk 6 above)
Comment (MBkkt) [reply core/formats/columnstore.cpp:2015]: I mean in your code we want to know that kNormal is zero and ColumnHint values not intersect

File: core/index/index_writer.hpp [hunk 16]
```diff
@@ -202,11 +202,12 @@ class index_writer : private util::noncopyable {
     /// @brief create a document to filled by the caller
     ///        for insertion into the index index
     ///        applied upon return value deallocation
+    /// @arg no_flush don't trigger segment flush
     /// @note the changes are not visible until commit()
     ////////////////////////////////////////////////////////////////////////////
-    document insert() {
+    document insert(bool no_flush = false) {
```
Comment (MBkkt) [inline core/index/index_writer.hpp:208]: Maybe `flush` or `need_flush` instead of `no_flush`

File: core/utils/attribute_helper.hpp (hunk 1 above)
Comment (gnusi) [reply core/utils/attribute_helper.hpp:58]: I think that's due to `std::get<I>(t)` might return either `T&` or `attribute_ptr<T>&`.

File: core/index/index_writer.hpp [hunk 17]
```diff
@@ -225,7 +226,7 @@ class index_writer : private util::noncopyable {
     void remove(Filter&& filter) {
       // thread-safe to use ctx_/segment_ while have lock since active flush_context will not change
       // cppcheck-suppress unreadVariable
-      auto ctx = update_segment(); // updates 'segment_' and 'ctx_'
+      auto ctx = update_segment(false); // updates 'segment_' and 'ctx_'
```
Comment (MBkkt) [inline core/index/index_writer.hpp:229]: also don't understand why you write default value

**Review by Dronplane — APPROVED** [review]
LGTM

File: core/formats/sparse_bitmap.cpp (hunk 11 above)
Comment (gnusi) [reply core/formats/sparse_bitmap.cpp:464]: Done

File: core/formats/columnstore2.cpp (hunk 9 above)
Comment (gnusi) [reply core/formats/columnstore2.cpp:1249]: Done

File: core/formats/columnstore2.cpp (hunk 8 above)
Comment (gnusi) [reply core/formats/columnstore2.cpp]: Done

File: core/formats/sparse_bitmap.cpp (hunk 14 above)
Comment (gnusi) [reply core/formats/sparse_bitmap.cpp:43]: Done

File: core/index/column_info.hpp [hunk 18]
```diff
@@ -23,49 +23,22 @@
...
- private:
-  type_info compression_;
-  compression::options options_;
-  bool encryption_;
-}; // column_info
+struct column_info {
+  type_info compression{irs::type<irs::compression::none>::get()};
+  compression::options options{};
+  bool encryption{};
+  bool track_prev_doc{};
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/index/column_info.hpp:37]: Done

File: core/formats/sparse_bitmap.hpp (hunk 15 above)
Comment (gnusi) [reply core/formats/sparse_bitmap.hpp:184]: I think the implementation differs from the default behavior.

Comment (gnusi) [reply core/formats/sparse_bitmap.hpp:184]: Done

File: core/formats/columnstore.cpp (hunk 6 above)
Comment (gnusi) [reply core/formats/columnstore.cpp:2015]: Done

File: core/formats/columnstore.cpp (hunk 7 above)
Comment (gnusi) [reply core/formats/columnstore.cpp:2161]: Done

File: core/formats/columnstore.cpp (hunk 6 above)
Comment (gnusi) [reply core/formats/columnstore.cpp:2015]: Done

File: core/formats/columnstore.cpp (hunk 7 above)
Comment (gnusi) [reply core/formats/columnstore.cpp:2161]: Done

File: core/index/index_writer.hpp (hunk 16 above)
Comment (gnusi) [reply core/index/index_writer.hpp:208]: done

File: core/formats/sparse_bitmap.cpp (hunk 4 above)
Comment (gnusi) [reply core/formats/sparse_bitmap.cpp:381]: It's already written in BE

File: core/formats/sparse_bitmap.cpp (hunk 13 above)
Comment (gnusi) [reply core/formats/sparse_bitmap.cpp:520]: Done

File: core/formats/sparse_bitmap.cpp (hunk 12 above)
Comment (gnusi) [reply core/formats/sparse_bitmap.cpp:486]: Done

**Pull request merged**